coil-core/src/androidMain/kotlin/coil3/decode/StaticImageDecoder.kt

```diff
@@ -106,40 +108,45 @@ class StaticImageDecoder(
             options: Options,
             imageLoader: ImageLoader,
         ): Decoder? {
-            val source = result.source.imageDecoderSourceOrNull(options) ?: return null
+            val source = result.source.toImageDecoderSource(options) ?: return null
             return StaticImageDecoder(source, result.source, options, parallelismLock)
         }
+    }
+}
 
-        private fun ImageSource.imageDecoderSourceOrNull(options: Options): ImageDecoder.Source? {
-            if (fileSystem === FileSystem.SYSTEM) {
-                val file = fileOrNull()
-                if (file != null) {
-                    return ImageDecoder.createSource(file.toFile())
-                }
-            }
+@InternalCoilApi
+@RequiresApi(28)
+fun ImageSource.toImageDecoderSource(options: Options): ImageDecoder.Source? {
+    if (fileSystem === FileSystem.SYSTEM) {
+        val file = fileOrNull()
+        if (file != null) {
+            return ImageDecoder.createSource(file.toFile())
+        }
+    }
 
-            val metadata = metadata
-            return when {
-                metadata is AssetMetadata -> {
-                    ImageDecoder.createSource(options.context.assets, metadata.filePath)
-                }
-                metadata is ContentMetadata -> try {
+    val metadata = metadata
+    return when {
+        metadata is AssetMetadata -> {
+            ImageDecoder.createSource(options.context.assets, metadata.filePath)
+        }
+        metadata is ContentMetadata -> run {
+            if (SDK_INT >= 29) {
+                try {
                     // Ensure the file descriptor supports lseek.
                     // https://github.com/coil-kt/coil/issues/2434
                     val asset = metadata.assetFileDescriptor
                     Os.lseek(asset.fileDescriptor, asset.startOffset, SEEK_SET)
-                    ImageDecoder.createSource { asset }
-                } catch (_: ErrnoException) {
-                    ImageDecoder.createSource(options.context.contentResolver, metadata.uri.toAndroidUri())
-                }
-                metadata is ResourceMetadata && metadata.packageName == options.context.packageName -> {
-                    ImageDecoder.createSource(options.context.resources, metadata.resId)
-                }
-                metadata is ByteBufferMetadata -> {
-                    ImageDecoder.createSource(metadata.byteBuffer)
-                }
-                else -> null
+                    return@run ImageDecoder.createSource { asset }
+                } catch (_: ErrnoException) {}
             }
+            ImageDecoder.createSource(options.context.contentResolver, metadata.uri.toAndroidUri())
+        }
+        metadata is ResourceMetadata && metadata.packageName == options.context.packageName -> {
+            ImageDecoder.createSource(options.context.resources, metadata.resId)
+        }
+        metadata is ByteBufferMetadata && (SDK_INT >= 30 || metadata.byteBuffer.isDirect) -> {
```

Comment:
For static image on SDK 29, 30
decode a `HeapByteBuffer` is safe as ImageDecoder will decode it immediately.